### ISSUE TITLE
swarm: Managers back off to Worker role without demoting

### ISSUE THREAD

**Issue by abronan** [docker/docker#24651]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-dev
 API version:  1.25
 Go version:   go1.6.2
 Git commit:   56fb100
 Built:        Thu Jul 14 18:01:59 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-dev
 API version:  1.25
 Go version:   go1.6.2
 Git commit:   56fb100
 Built:        Thu Jul 14 18:01:59 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.12.0-dev
Storage Driver: overlay
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: null host bridge overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-28-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 1.954 GiB
Name: node01
ID: ZTNM:55LA:E6YV:QC3Y:5QNC:LFMK:L4IE:MQ3Q:S43M:D45D:H4P7:HPLC
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 15
 Goroutines: 23
 System Time: 2016-07-14T16:45:45.518075087-04:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Digital Ocean VMs:

*3 Managers*: 2 GB Memory / 40 GB Disk / Ubuntu 16.04 x64
*3 Agents*: 1 GB Memory / 30 GB Disk / Ubuntu 16.04 x64

**Steps to reproduce the issue:**

1. Create the cluster with 3 Managers and 3 Agents
2. Create a service
3. Scale the service to a high value (1000/2000)

Output of `docker node ls`:
```
ID                           HOSTNAME  MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
0kcj1xdtbngheh54s1dxitevi    node06    Accepted    Ready   Active        
0zy1wr8fh9lbs4myosr5pqhmx    node02    Accepted    Ready   Active        Reachable
124kqfd9u7eo3w1qsdfmrr27s    node05    Accepted    Ready   Active        
7bpe7qdm50o7k15m3vormjd12    node04    Accepted    Ready   Active        
anbi7afpj7aoj5lj4nnffoub1    node03    Accepted    Ready   Active        Reachable
e3lsnbu80kepinnzriq0gwc9m *  node01    Accepted    Ready   Active        Leader
```

**Describe the results you received:**

The Managers will all drop their Manager status and fall back to Worker mode somehow giving:

```
Error response from daemon: This node is not a Swarm manager. Worker nodes can't be used to view or modify cluster state. Please run this command on a manager node or promote the current node to a manager.
```

on `docker node ls` for the 3 Managers.

Raft is still running in the background but seems like another component is causing the Manager to stop in a weird way (there are error logs to state that the dispatcher is stopped at this point).

**Describe the results you expected:**

Manager keeps running and working even in degraded scenario (like this was the case recently). I'm not sure if this is a problem with certificate issuance or role assignment. But the Managers back off to Worker role even though some of the components are still up and running (like raft).

**Additional information you deem important (e.g. issue happens only occasionally):**

Seems to happen consistently. Will try to navigate through the flow of logs to see what happened exactly.

**Comment by cpuguy83** [comment]
Any chance you have logs from this? Goroutine crash?

**Comment by abronan** [comment]
No crash now that the daemon has a lower default OOM score, but the sum of operations going on due to the massive scale operation makes it very hard to parse what is going on.

After running `pprof`, it seems that we spend most of the time swapping memory to disk, thus there might be some issues with certificate issuance which makes the daemon "think" it is back to Worker mode after some time.

Somehow I didn't encounter this behavior while testing #24516  

Also I've noticed that if I scale down just before the Managers back off to Worker mode, they are in `Worker` mode for some time but go back to `Manager` once the cluster stabilizes and probably has more resources thanks to the scale down operation.

I'll retry with current `master` and see if I can reproduce.

**Comment by dongluochen** [comment]
I tried to run 1000 redis on 5 node cluster with command `docker service create --replicas 1000 redis:3.0.7`. I lose access to cluster after that. 

```
ubuntu@ip-172-19-241-147:~$ docker node ls
Error response from daemon: rpc error: code = 4 desc = context deadline exceeded
ubuntu@ip-172-19-241-147:~$ docker service ls
Error response from daemon: rpc error: code = 4 desc = context deadline exceeded
ubuntu@ip-172-19-241-147:~$ docker ps -a | wc -l
125
```

Raft reports `raft: failed to process the request: node lost leader status` after this service.

```
grep -i "Jul 19 23:" backuplog | grep -i leader
Jul 19 23:02:54 ip-172-19-241-145 dockerd[8292]: time="2016-07-19T23:02:54.030834384Z" level
=info msg="raft.node: 2cf0d4a8eb0185fa lost leader 57565e9aad5b2018 at term 6"
Jul 19 23:03:00 ip-172-19-241-145 dockerd[8292]: time="2016-07-19T23:03:00.144399970Z" level
=info msg="2cf0d4a8eb0185fa became leader at term 7"
Jul 19 23:03:00 ip-172-19-241-145 dockerd[8292]: time="2016-07-19T23:03:00.144580719Z" level
=info msg="raft.node: 2cf0d4a8eb0185fa elected leader 2cf0d4a8eb0185fa at term 7"
Jul 19 23:05:08 ip-172-19-241-145 dockerd[8292]: time="2016-07-19T23:05:08.954771779Z" level
=info msg="raft.node: 2cf0d4a8eb0185fa lost leader 2cf0d4a8eb0185fa at term 35"
Jul 19 23:05:09 ip-172-19-241-145 dockerd[8292]: time="2016-07-19T23:05:09.006052246Z" level
=error msg="failed to move all nodes to \"unknown\" state: raft: failed to process the reque
st: node lost leader status" module=dispatcher
Jul 19 23:05:09 ip-172-19-241-145 dockerd[8292]: time="2016-07-19T23:05:09.149576141Z" level
=error msg="failed allocating ingress network during init: raft: failed to process the reque
st: node lost leader status"
Jul 19 23:05:09 ip-172-19-241-145 dockerd[8292]: time="2016-07-19T23:05:09.496159301Z" level
=error msg="scheduler tick transaction failed" error="raft: failed to process the request: n
ode lost leader status"
```

**Comment by abronan** [comment]
@dongluochen This makes sense as scheduling that many tasks exhausts memory, and we are stuck in a loop where the OOM trigger kills tasks but the orchestrator keeps retrying. Most of the time is spent swapping and the cluster becomes completely blocked. Gossip and Raft heartbeats deadlines are missed, vote requests are ignored, etc.

While the cluster becomes unstable in such scenario, Managers should keep their role until we are able to fix the problem with a `docker service scale redis=50` for example. In my case the Managers are literally dropping their role and they consider themselves as Workers after some time. Which makes it impossible to fix the problem.

Restarting the Managers fixes the problem so there is a short time window where this can be fixed with scaling down but I would expect to be able to avoid this issue without restarting the daemons.

The real solution is documentation I think and using `--availability drain` on the Managers. Yet as Manager/Worker is the default, It would make sense to avoid such a scenario to happen in the first place.

I'm not sure what a good solution could be, maybe reserving a fixed chunk of memory for the daemon that no container could use if this is something that could be achieved. But this is a complex problem that a critical piece like Raft runs alongside containers without exclusive resource constraints.

**Comment by dongluochen** [comment]
@abronan In my case the cluster never recovered. Restarting managers doesn't help. I have to re-bootstrap the cluster. 

I think service scaling may take the approach of service rolling update. New tasks are created in batches. Users can stop the rollout and change the service. It'd prevent fat finger errors.

**Comment by abronan** [comment]
@dongluochen I don't think anything will prevent this issue from happening as long as raft/gossip and other critical components don't have reserved resources that other process can't use/override.

I opened docker/swarmkit#1156 to track the service scaling potential user error.

Also opened docker/docker#24857 to recommend setting the Managers with `--availability drain` for scenarios where users might scale containers without resource constraints.

**Comment by thaJeztah** [comment]
@abronan should I change this to a "documentation" issue?

/cc @sfsmithcha

**Comment by thaJeztah** [comment]
Actually; I see we have the documentation merged; I'm closing this issue because https://github.com/docker/docker/pull/24857 is merged
